veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,330 @@
+```
+
+Once the raw restored image is created, you can store it in a PVC and use this PVC as the restored volume for the VM. You can use any population method like import or upload.
+
+
+## Alternatives
+
+The option to use [KEP-3314: CSI Changed Block Tracking](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/3314-csi-changed-block-tracking), the native Kubernetes approach for CBT, was considered as an alternative to facilitating libvirt backups.
+Advantages:
+- Aligning with k8s ecosystem.
+- Avoid the need to implement this complex feature ourselves.
+- Backup vendors already use CSI API currently for full backups.
+
+Disadvantages:
+- This KEP is still in the early stages of implementation and will take considerable time to mature before reaching a stable v1 version and it is not under our control.
+- Beyond waiting for the Kubernetes API to mature, each storage provider must implement the optional SnapshotMetadata API as outlined in the KEP. Since this is optional, not all providers will support it, and adoption will be gradual, restricting the available options for users.
+- When using Kubernetes VolumeSnapshot, the guest must be frozen during the snapshot of all the VM volumes. In contrast, libvirt uses a dedicated job to capture the disk state quickly, minimizing guest downtime.
+- The API is limited to the design choices made by Kubernetes, offering little flexibility for adjustments or additions.
+
+## Scalability
+
+- PVC to contain the scratch backup space which saves any blocks that change during the backup job, this PVC can in worst case scenario can grow up to the size of the sum of all disks size.
+- QCOW2 overlay requires a minimal PVC to be stored in. With certain storage providers even when requested small PVC size they provide a larger one according to their minimal PVC size. (This is a general issue with VM state PVCs)
+
+## Update/Rollback Compatibility
+
+Since the new feature allows users to enable or disable it, upgrades will not pose any issues. Users must opt in by setting the QCOW2Overlay to true. The rollback will not be a problem either, as it is essentially the same as setting the QCOW2Overlay to false, which will be the default value.
+
+## Functional Testing Approach
+
+Testing should check data consistency before and after add and remove of the QCOW2 overlay.
+Data consistency of incremental backup.
+Data consistency after VM restart.
+Check failure scenario where incremental backup cannot be done and in such case full backup should be required.
+
+## Implementation Phases
+
+- Add/remove the qcow2 overlay
+- Subresource to initiate backup (full and incremental) including Libvirt wrapper backup functions.
+- New VirtualMachineBackup CR + controller for backups - online backup only
+- Handle restart of VM and redefinition of checkpoints.
+- Offline backup
+- API to allow to pull the backup.
+- Handle VM failure where bitmap is not valid - next backup needs to be full.
```

Comment:
This should be done earlier in my opinion.  Probably during the phase where you handle VM restart and redefinition of the checkpoints since this is the logical place where we will check for dirty bitmap consistency.

Reply:
sure make sense